### PR TITLE
Fix Issue 13649 - uniform01 Assertion failure

### PR DESCRIPTION
**Pull request by e10s**
https://issues.dlang.org/show_bug.cgi?id=13649

We have to throw away the result and try again when it is rounded to 1. As a result, this PR disables 93fe844.

See also [boost/random/uniform_01](http://www.boost.org/doc/libs/1_57_0/boost/random/uniform_01.hpp), which is the original, according to #2050.

### CODE REVIEW

**Comment by quickfur** [comment]
Ah I see, `int.max - int.min` overflows float, so the assumption that `isIntegral` implies strict inequality fails to hold.

LGTM.

**Comment by monarchdodra** [comment]
Ping @WebDrake

**Comment by quickfur** [comment]
I think this looks good to merge, if @WebDrake doesn't respond. It just disables a previous, apparently not-completely-correct, optimization; it doesn't change the semantics AFAICT.

**Comment by WebDrake** [comment]
What's worrying me is that simply disabling the optimization may be papering over the cracks: if `rng.max - rng.min` overflows the floating-point type, can we guarantee the uniformity of the distribution in any case?

**Comment by quickfur** [comment]
Good point. So what should be done in that case?

**Comment by WebDrake** [comment]
BTW just for reference, in phobos, the RNG element types are always unsigned integers (this accords with the C++11 concept of a random number generator).

**Comment by WebDrake** [comment]
Well, one option would be to statically enforce that the floating-point type has enough precision to not overflow when handling the unsigned integer values.  That's a little disappointing because it makes uniform01!float impossible, but a tolerable solution, I think, given that it's obviously never worked in the first place.

Probably better would be to statically check the floating point type has enough precision: if so, you can apply the optimization; if not, maybe a `pragma(msg, ...)` or some other kind of compile-time warning to alert the user that they're using insufficient precision, while using the runtime check on `u < 1`.

**Comment by WebDrake** [comment]
For reference:

```D
float u = (uint.max - uint.min); // evaluates to 4294967296
uint u2 = (uint.max - uint.min); // evaluates to 4294967295

**Comment by WebDrake** [comment]
The above is potentially somewhat reassuring inasmuch as it means that reverting to runtime check is probably going to be OK, but I'm slightly mystified why the transition to `float` should evaluate to exactly `(uint.max - uint.min) + 1`.

**Comment by monarchdodra** [comment]
Because there isn't enough precision...? That's the root cause of the
issue. Since you are "missing" bits of precision, float has to round to
2^^howmanybitsyouaremissing.

BTW, would using doubles, but casting back down to floats after wards be an
acceptable solution?

That said, we'd still have to check vs long.

**Comment by WebDrake** [comment]
Yea, of course -- I'm being braindead after a too-long day :-)  It's rounding to nearest power of 2 and that happens to be the one (immediately) above the actual integral value.  The next number down, for `float`, is not `2 ^^ 32 - 1` but `2 ^^ 32 - 256` ... !

I think this problem will recur for `double` if the RNG element type is `ulong`, no ... ?

Anyway, I suggest that the constraint be replaced with something like,

```D
static if (isIntegral!R && T.sizeof > R.sizeof) { ... }
```

... which should be commented internally to explain why.

Ideally I suppose you'd check using the significand of the floating-point type, but that's probably overkill, as the above will work for all built-in types, I think ... ?

**Comment by WebDrake** [comment]
BTW this might be overkill, but a ready way to unittest the function for future reference, is to create a mock "Dilbert-esque" fake-RNG, templated on a word `W = ushort, uint, ulong`, that theoretically has a range from `W.min` to `W.max` but in fact always returns `W.max`.  Then you can do something like:

```D
foreach (W; TypeTuple!(ushort, uint, ulong))
{
    AlwaysMax!W fake_rng;    // .front always returns W.max
    foreach (T; TypeTuple!(float, double, real))
    {
        T u = uniform01!T(fake_rng);
    }
}
```

**Comment by monarchdodra** [comment]
Sounds like a job for a generic adaptor? BTW, don't we have a PRNG that can generate numbers from a finite set? It seems both `randomSample` and `randomCover` return finite ranges. I'd have suggested something like `dice([0.5, 0.5]).map!(a => [0u, uint.max][a])()`, zhich is infinite and random, but is not actually a PRNG, as it does not have the prng enum (which kind of brings up the question about how lenient we want to be about the `isPrng` definition...). Or just `cycle([0u, uint.max])`

Just saying I'm pretty sure we can do MUCH better than `alwaysMax`.

Here's an idea: `toPrng(infinteInputRange)`: Takes a range and returns the same range, but that satisfies `isPrng`. DONE! The keyword of course here is the "P" of "Pseudo", as the resulting range could be "not that random..."

Anyways, with `toPrng` and generic ranges, you should have more than enough tools to do any testing you want I believe?

**Comment by monarchdodra** [comment]
Also, obligatory xkcd Random Number
![RFC 1149.5 specifies 4 as the standard IEEE-vetted random number.]
(http://imgs.xkcd.com/comics/random_number.png)

http://imgs.xkcd.com/comics/random_number.png

File: std/random.d [hunk 1]
```diff
@@ -1726,24 +1726,16 @@ body
     {
         immutable T u = (rng.front - rng.min) * factor;
         rng.popFront();
-        static if (isIntegral!R)
+
+        /* If RNG variates are integral, u looks always less than 1 in view of
+         * the definition of factor. However, we must consider floating-point
+         * calculation errors. Moreover, we have to check if a floating-point
```
Comment (don-clugston-sociomantic) [inline std/random.d]: What does "we must consider floating-point calculation errors" mean?
Perhaps you mean, "Although the infinitely precise result is strictly less than 1.0, it will be rounded up to 1.0 if is greater than 1 - T.epsilon/2".

**Comment by don-clugston-sociomantic** [comment]
Even if this is integral, what happens if  UniformRNG returns signed integers over the full bit pattern (which you would hope it does, in the case where it returns signed integers)?
Then if   ``rng.min == - (int.max - 1)`` and ``rng.max = int.max``, so
``factor = 1 / (1 - rng.max - rng.min)``  is a division by zero. Right?
I don't understand how this function can be correct.  And the boost function doesn't inspire confidence with this comment in the code:
``  * Note: The current implementation is buggy, because it may not fill all of the mantissa with random bits.``

**Comment by don-clugston-sociomantic** [comment]
Honestly I'm shocked that boost accepted code like that without comments. Why is 1 added to the factor in the integral case? It's not at all obvious to me that the code is correct. The mathematics is quite subtle, it is very easy to get this sort of stuff wrong.

File: std/random.d (hunk 1 above)
Comment (e10s) [inline std/random.d]: Ya, exactly.

**Comment by monarchdodra** [comment]
> Then if rng.min == - (int.max - 1) and rng.max = int.max, so

No, it's `rng.min == - (int.max + 1)`. Rule of thumb is that there are the same amount of positive/negative numbers, but the positives start at 0, whereas negatives start at -1. So the most negative is the most positive + 1. Eg byte: [-128, 127].

> factor = 1 / (1 - rng.max - rng.min)

Not sure where that formula comes from. Typo? The formular used is `1/(T(1) + (rng.max - rng.min))`. In any case, no, it's not 0. Furthermore, the factor is pretty much unchanged regardless of the signed-ness of `R`.

> Honestly I'm shocked that boost accepted code like that without comments. Why is 1 added to the factor in the integral case?

The "1" is there in the sense that "the ammount of values between A and B *inclusive* is B - A *+1* ".

So in this case, if a PRNG can generate values between min and max (inclusive), it can generate `1 + max - min` different values, hence the factor.

For floating point types, the formula is different because (AFAIK, there was a discussion about this, do you remember it @WebDrake ?) there is an unwritten assumption that a floating point PRNG will never produce it's max.

TL;DR:
factor = 1/range
Integral PRNG's use `"[]"` ergo range = diff + 1
Floating PRNG's use `"[)"` ergo range = diff

**Comment by monarchdodra** [comment]
> It's not at all obvious to me that the code is correct. The mathematics is quite subtle, it is very easy to get this sort of stuff wrong.

Yes, agreed this kind of stuff needs special attention.

**Comment by don-clugston-sociomantic** [comment]
Yeah, I mistyped the formula in both cases. Consider the byte case.  
``1/(T(1) + (rng.max - rng.min)) == 1 / ( 1 + (127 - (-128))  =  1 /  256``.  Note that the denominator (256) does not fit in one byte.  So the T(1) instead of 1 is crucial to avoid division by zero. That's really subtle.

I can't see how this can produce uniform random numbers if the range rng.min.. rng.max is not a power of 2. (Ie, the code is not a generic as it pretends to be).

**Comment by monarchdodra** [comment]
> I can't see how this can produce uniform random numbers if the range rng.min.. rng.max is not a power of 2. (Ie, the code is not a generic as it pretends to be).

Actually, I'm not sure how the power of 2 is relevant? Perhaps you mean if it doesn't provide full 32 bits?

What it provides is a uniform distribution over [0..1), with the *granularity* provided by the passed prng. Say, if your PRNG can only provide [0, 1, 2], then you will generate uniform numbers over [0..1), "within" the universe of [0, 0.33, 0.66].

It's no different from generating doubles with a uint generator: you will only have 32 bits of mantissa granularity.

Though this should arguably be documented.

**Comment by monarchdodra** [comment]
BTW, I believe @WebDrake 's suggestion of `static if (isIntegral!R && T.sizeof > R.sizeof)` is correct and should be used.

**Comment by e10s** [comment]
I haven't understood why `T.sizeof > R.sizeof` is reasonable yet. Any simple examples?

**Comment by monarchdodra** [comment]
> I haven't understood why T.sizeof > R.sizeof is reasonable yet. Any simple examples?

It's a simple "does my floating point type have enough mantissa bits to hold an integral type without loss of precision". Here's the truth table (sign irrelevant):

```
       L I S B (long int short byte)
double N Y Y Y
float  N N Y Y
```

The "simplest" way to "express" this truth table is to notice that it also illustrates the rule for `Float.sizeof > Integral.sizeof`.

**Comment by e10s** [comment]
Okay, I got it. Anyway, it seems a better way to avoid unexpected results!

**Comment by WebDrake** [comment]
@don-clugston-sociomantic 
> Even if this is integral, what happens if UniformRNG returns signed integers over the full bit pattern

There's a tacit assumption in phobos (which explicitly specified in section 26.5.1.3 of the C++11 standard, from which `std.random`'s design broadly derives) that uniform RNGs return unsigned integer types.  This should quite possibly be made a strict requirement (it is in hap.random: https://github.com/WebDrake/hap/blob/master/source/hap/random/traits.d#L47).

> the T(1) instead of 1 is crucial to avoid division by zero. That's really subtle.

Yup, I really ought to have documented that when I wrote the original function.  IIRC I ran into the division-by-zero issue when drafting it and realized why boost did it this way.

@monarchdodra 
> TL;DR:
> factor = 1/range
> Integral PRNG's use "[]" ergo range = diff + 1
> Floating PRNG's use "[)" ergo range = diff

Yes, indeed.  For example, Boost.Random documentation describes this distinction, and the GSL has some double-producing random number generators that also operate this way.  That said, I would never be comfortable making an absolute assumption that all floating-point RNGs would have an open upper bound.

> I can't see how this can produce uniform random numbers if the range rng.min.. rng.max is not a power of 2. (Ie, the code is not a generic as it pretends to be).

I do wonder if that might be why the C++11 standard makes the specifications it does for uniform RNGs using unsigned integer types.  There are in any case subtleties to do with the slight difference in bit precision of integral and floating-point types.  However in the case of `uniform01` this is arguably forgiveable as the whole point is to produce the variate really fast with "good enough" uniformity.  This should be better documented (my bad).

**Comment by WebDrake** [comment]
@don-clugston-sociomantic out of curiosity, is there a(n easy) way to verify if the significand of a floating-point number has enough bits to represent the full range of an integral type?

**Comment by WebDrake** [comment]
@klickverbot works for floating-point but what about the integral type?  I don't want to make a hacky assumption that just assumes 8 bits per byte.  `CHAR_BIT` from `core.stdc.limits` ... ?

**Comment by WebDrake** [comment]
IOW we could do something like

```D
import core.stdc.limits: CHAR_BIT;
static if (isIntegral!R && T.mant_dig >= (CHAR_BIT * R.sizeof))
{
    ... 
}
```

**Comment by monarchdodra** [comment]
> I don't want to make a hacky assumption that just assumes 8 bits per byte. `CHAR_BIT` from `core.stdc.limits` ... ?

Unlike C (and C++), D specs MANDATE madates the sizes of all its primitives, as well as the number of bits in a byte (8). So using a raw `8` is perfectly fine and not hacky at all.

You can still import `core.stdc.limits` if you feel a name over a magic number is better, but it adds a not so needed dependency. AFAIK, this magic number does not have any built-in alias. I'd love to be proven wrong though.

BTW, `T.mant_dig >= (CHAR_BIT * R.sizeof)` is a more expressive check (IMO). I like it.

**Comment by don-clugston-sociomantic** [comment]
> What it provides is a uniform distribution over [0..1), with the granularity provided by the passed prng. Say, if your PRNG can only provide [0, 1, 2], then you will generate uniform numbers over [0..1), "within" the universe of [0, 0.33, 0.66].  

The problem is that the uniformity is broken. It's wrong in much the same way that creating a dice roll by ``rand() % 6`` is wrong - not all numbers are equally likely.

For a similar reason I'm surprised by the default floating rng range being "[ )".  It's odd because open ranges don't really exist in floating point numbers in the default rounding mode. Once you consider rounding, an open range is identical to a closed range with the same limits. Actually when we're saying ``[0 .. 1)``we really mean ``[0..nextDown(1)]``. But that's really correct only in the "round down" rounding mode. If the distribution was really uniform, then the probability of returning 1 would be half the probability of nextDown(1.0). So we're going to extra trouble to make the distribution slightly non-uniform!

What I think it should be doing is: calculate the number of bits via ``mant_dig``, and divide this by the number of bits in the uniform source.  If we want to avoid ever returning 1, then for the last few bits we should mask them out of the RNG. 
Eg for double there are 52 bits of mantissa, if the RNG returns 32 bits we want:
``((rand & 0xFFFF_F000_0000) / 2.0 ^^32) + rand ) / 2.0^^32``
(noting that 0xFFFF_F000_0000 has 20 bits set).
So we have something like (untested):
```
enum int chunks_required = T.mant_dig / ( R.sizeof * 8);
 enum R remainder = T.man_dig - (R.sizeof *8 * chunks_required);
enum  factor = 1.0 / (R.sizeof * 8);
static if (remainder > 0) 
{
    // value of the form 0b11110000000, where number of set bits is equal to the number of remainder bits
    enum remainder_bitmask = ((1 << (R.sizeof *8)) -1 ) - ((1 << R.sizeof *8 - remainder ) - 1);
   T val = (rng.front & remainder_bitmask) * factor;
    rng.popFront();
}
else
{
   T val = 0;
}
static if (chunks_required > 0) {
   val = (val +  rng.front()) * factor;
   rng.popFront();
}
static if (chunks_required > 1) {
   val = (val +  rng.front()) * factor;
   rng.popFront();
}
// etc, up to the maximum supported mant_dig, we could really use static foreach here!

return val;
```

**Comment by don-clugston-sociomantic** [comment]
And BTW the reason why it is normally an open interval, is probably so that you can just poke the random bits directly into the mantissa. Failing to ever return 1 is probably a side-effect, not a desirable feature.

**Comment by monarchdodra** [comment]
> The problem is that the uniformity is broken. It's wrong in much the same way that creating a dice roll by rand() % 6 is wrong - not all numbers are equally likely.

How so? In my example, `[0, 0.33, 0.66]` ARE all equally likely. (unlike, indeed, the `% 6` arithmetic)

> What I think it should be doing is...

didn't look at all the details, but sounds right.

**Comment by WebDrake** [comment]
> The problem is that the uniformity is broken. It's wrong in much the same way that creating a dice roll by `rand() % 6` is wrong - not all numbers are equally likely.

It's not as bad as _that_!  (Random numbers by modulo is one of my pet peeves.)

It's worth reiterating that `uniform01` is, by design, a tradeoff in favour of speed over precision of randomness -- one that's good enough for a great many use cases.  However, it would great if we can get the speed without the cost, which your draft version suggests might be possible.

I'll need to look at it in more depth, and it probably won't be 'til next week, but I'll try and make the time to really test this out.

**Comment by e10s** [comment]
Updated to WebDrake's `CHAR_BIT` version.

**Comment by WebDrake** [comment]
LGTM.  Can you squash the two patches together please?

**Comment by e10s** [comment]
Rebased!

**Comment by monarchdodra** [comment]
LGTM. Guys, any outstanding objections?

**Comment by monarchdodra** [comment]
Well, those who are absent, are always wrong. Merging.

**Comment by monarchdodra** [comment]
Auto-merge toggled on
